File: Ruby/parens.rb
#RUBY - Generate Valid Combinations of n Parentheses pairs
# Given an integer, generate an array containing all pairs of valid parentheses.
# Parentheses are valid if (()()(()))() each pair opens and closes properly. For example, )( would NOT be valid! (Neither would ()))
# generateParens(1) => ["()"]
# generateParens(2) => ["()()","(())"]
# generateParens(3) => ["()()()", "()(())", "(())()", "(()())", "((()))"]


## STRING Solution
def generateParensString(n)
  if n > 0
    result =  _generateParensString(n, "", [], 0, 0)
    print "input of #{n}: " + result.to_s + "\n"
  end
end

def _generateParensString(n, item, valid, open, close)
  if close == n  # count of closed gets to n
    valid.push(item)
    return
  else
    if open > close
      updatedItem = item + ")"
      _generateParensString(n, updatedItem, valid, open, close+1)
    end
    if open < n
      updatedItem = item + "("
      _generateParensString(n, updatedItem, valid, open+1, close)
    end
  end
  return valid
end

## ARRAY solution
def generateParensArray(n)
  if n > 0
    result = _generateParensArray(n, [], [], 0, 0, 0)
    print "input of #{n}: " + result.to_s + "\n\n"
  end
end

def _generateParensArray(n, array, valid, i, open, close)
  if close == n # count of closed gets to n
    current = array.join("").to_s
    valid.push(current)
  else
    if open > close
      array[i] = ")"
      _generateParensArray(n, array, valid, i+1, open, close + 1)
    end
    if open < n
      array[i] = "("
      _generateParensArray(n, array, valid, i+1, open + 1, close)
    end
  end
  return valid
end

puts "\n- - - - - - String - - - - -\n"
generateParensString(2)
puts "\n- - - - - - Array - - - - - -\n"
generateParensArray(3)

File: GO/parens.go
package main

import "fmt"

// GO - Generate Valid Combinations of n Parentheses pairs
// Given an integer, generate an array containing all pairs of valid parentheses.
// Parentheses are valid if (()()(()))() each pair opens and closes properly. For example, )( would NOT be valid! (Neither would ()))
// generateParens(1) => ["()"]
// generateParens(2) => ["()()","(())"]
// generateParens(3) => ["()()()", "()(())", "(())()", "(()())", "((()))"]

// Strings - one at a time
func recursiveParens(n int, current string, open int, close int) {
	if close == n {
		fmt.Println(current)
		return
	} else {
		if open > close {
			newCurrent := current + ")"
			recursiveParens(n, newCurrent, open, close+1)
		}
		if open < n {
			newCurrent := current + "("
			recursiveParens(n, newCurrent, open+1, close)
		}
	}
	return
}

// Array of strings
func recursiveParensArray(n int, current string, results *[]string, open int, close int) *[]string {
	if close == n {
		fmt.Println(current)
		*results = append(*results, current)


		return results
	} else {
		if open > close {
			newCurrent := current + ")"
			recursiveParensArray(n, newCurrent, results, open, close+1)
		}
		if open < n {
			newCurrent := current + "("
			recursiveParensArray(n, newCurrent, results, open+1, close)
		}
	}
	fmt.Println(len(*results))
	return results
}

func main() {
	var myResults = &[]string{}
	fmt.Println("\n-----print parens-----\n")
	recursiveParens(2, "", 0, 0)
	fmt.Println("\n-----print array of parens-----\n")
	fmt.Println(recursiveParensArray(3, "", myResults, 0, 0 ))
}

